ipld/resolver/src/service.rs

```diff
@@ -322,20 +334,78 @@ impl<P: StoreParams> Service<P> {
         match event {
             content::Event::Complete(query_id, result) => {
                 if let Some(query) = self.queries.remove(&query_id) {
-                    self.handle_query_result(query, result);
+                    self.resolve_query(query, result);
                 } else {
                     warn!("query ID not found");
                 }
             }
         }
     }
 
+    /// Handle an internal request coming from a [`Client`].
+    fn handle_request(&mut self, request: Request) {
+        match request {
+            Request::SetProvidedSubnets(ids) => {
+                if let Err(e) = self.membership_mut().set_provided_subnets(ids) {
+                    warn!("failed to publish set provided subnets: {e}")
+                }
+            }
+            Request::AddProvidedSubnet(id) => {
+                if let Err(e) = self.membership_mut().add_provided_subnet(id) {
+                    warn!("failed to publish added provided subnet: {e}")
+                }
+            }
+            Request::RemoveProvidedSubnet(id) => {
+                if let Err(e) = self.membership_mut().remove_provided_subnet(id) {
+                    warn!("failed to publish removed provided subnet: {e}")
+                }
+            }
+            Request::PinSubnet(id) => self.membership_mut().pin_subnet(id),
+            Request::UnpinSubnet(id) => self.membership_mut().unpin_subnet(&id),
+
+            Request::Resolve(cid, subnet_id, response_channel) => {
+                self.start_query(cid, subnet_id, response_channel)
+            }
+        }
+    }
+
+    /// Start a CID resolution.
+    fn start_query(&mut self, cid: Cid, subnet_id: SubnetID, response_channel: ResponseChannel) {
+        let mut peers = self.membership_mut().providers_of_subnet(&subnet_id);
+
+        if peers.is_empty() {
+            send_resolve_result(response_channel, Err(anyhow!(NoKnownPeers(subnet_id))));
+        } else {
+            // Connect to them in a random order, so as not to overwhelm any specific peer.
+            peers.shuffle(&mut rand::thread_rng());
+
+            // Prioritize peers we already have an established connection with.
+            let (connected, known) = peers
+                .into_iter()
+                .partition::<Vec<_>, _>(|id| self.swarm.is_connected(id));
```

Comment:
sanity-check: partitioning maintains the order, right? If not we should probably shuffle after this.

Reply:
Great question! It does, it just loops through the input and appends to the two outputs. The unstable `partition_in_place` variant doesn't maintain the order. 

I actually started with shuffling both after, then concatenating, but didn't like to see all that code. So I thought: if we have a random list, and we partition it in a way that doesn't maintain the partial order of items, it's still shuffled, right?